server/services/draftsService.ts

```diff
@@ -0,0 +1,83 @@
+import { Request } from 'express'
+import uuid from 'uuid'
+
+export interface Draft<T> {
+  id: string
+  type: string
+  createdAt: Date
+  updatedAt: Date
+  data: T
+}
+
+interface DraftDTO {
+  id: string
+  type: string
+  createdAt: string
+  updatedAt: string
+  data: unknown
+}
+
+export interface Clock {
+  now: () => Date
+}
+
+export default class DraftsService {
+  constructor(private readonly clock: Clock) {}
+
+  private sessionKey(id: string): string {
+    return `draft:${id}`
+  }
+
+  private fetchDraftDTO(id: string, req: Request): DraftDTO | null {
+    const dto = req.session[this.sessionKey(id)]
```

Comment:
I see that we're using the RedisStore solution for session store. So I think it would be worthwhile testing this in an environment with multiple containers just to be certain its working as expected.

Looking at the config, I believe that the session store time-out is currently 120 minutes. Is this an expectable time-out for saving drafts? I think the users may expect their drafts to last a bit longer. Perhaps some wording to inform the users that their drafts will be saved for x minutes might be useful?

Reply:
Oh Sorry, I see that you made a comment about this in your documentation. 

You mention that it is 16 hours. How did you derive this?

From the code in `app.ts` I see: 

```
      cookie: { secure: config.https, sameSite: 'lax', maxAge: config.session.expiryMinutes * 60 * 1000 },
```

And the expiration defined in `config.ts`:

```
    expiryMinutes: Number(get('WEB_SESSION_TIMEOUT_IN_MINUTES', '120')),
```

I can't see `WEB_SESSION_TIMEOUT_IN_MINUTES` being set within the helm scripts so assuming it falls back to 120 minutes.

Reply:
Oh Sorry... I see that you made a comment about this in your documentation. 

I was looking at the timeout for the cookie rather than timeout for the session in redis store. You mention that it is 16 hours just curious where this is defined?

Reply:
You know what, now that you ask, I have no idea how I managed to arrive at 16 hours. The line that I've linked to in `app.ts` in the documentation clearly shows it to be 120 minutes, like you say. I'll update the documentation to reflect this.

As to whether it's an acceptable timeout, I'm not sure. It probably comes down to the length of the journeys that we're using this for.

An alternative would be to, instead of storing these drafts in the session, to store them centrally in Redis, keyed by the username of the logged-in user. That way they wouldn't expire with the session. We could set a long TTL on these, perhaps a few days, after which Redis would clean them up. Do you think that might be a better idea? I can make that change if so.

Reply:
> So I think it would be worthwhile testing this in an environment with multiple containers just to be certain its working as expected before deploying to production? Dev environment has 2 containers.

Yep, good idea. I was wondering if you know if any way to check which container a request gets routed to? (So that I can check it really is working with 2 containers and that it's not just that all the requests I test it with are coincidentally being routed to the same container.)

Reply:
> I was wondering if you know of any way to check which container a request gets routed to? 

We can simple do `kubectl log -f` on both containers. 
I suppose we can also test it locally by deleting the intervention-ui container and then starting it back up between requests.

Reply:
> An alternative would be to, instead of storing these drafts in the session, store them directly in Redis, keyed by the username of the logged-in user. That way they wouldn't expire with the session. 

My understanding is that the express session being backed with the Redis store will automatically store them directly in Redis straight away (otherwise we'd have the problem with multiple containers and may need to play around with the config)....

Reply:
I have another question (maybe for David) - when we redeploy, does the Redis container bounce as well? In which case all the session store will be lost given it's in-memory store.

Reply:
> We could set a long TTL on these, perhaps a few days, after which Redis would clean them up. Do you think that might be a better idea? I can make that change if so.

hmmm, I don't think it would be an issues setting it for a few days...

However, in order to reduce problems... we could inform the user that their draft will only be saved for x amount of minutes/hours. I don't know how to present that to the user though...

Reply:
Sorry to be polluting this comments ... another possible downside is that their session will not persist between browsers or different locations. As I think that the session store is based on the cookies stored in the browser.